
<a href="https://colab.research.google.com/github/kokchun/Machine-learning-AI22/blob/main/Exercises/E02_sklearn.ipynb" target="_parent"><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a> &nbsp; to see hints and answers.

---
# Scikit-learn exercises 

---
These are introductory exercises in Machine learning with focus in **scikit-learn** .

<p class = "alert alert-info" role="alert"><b>Note</b> that sometimes you don't get exactly the same answer as I get, but it doesn't neccessarily mean it is wrong. Could be some parameters, randomization, that we have different. Also very important is that in the future there won't be any answer sheets, use your skills in data analysis, mathematics and statistics to back up your work.</p>

<p class = "alert alert-info" role="alert"><b>Note</b> that in cases when you start to repeat code, try not to. Create functions to reuse code instead. </p>

<p class = "alert alert-info" role="alert"><b>Remember</b> to use <b>descriptive variable, function, index </b> and <b> column names</b> in order to get readable code </p>

The number of stars (\*), (\*\*), (\*\*\*) denotes the difficulty level of the task

---

## 0. EDA (*)

In the whole exercise, we will work with the "mpg" dataset from seaborn dataset. Start by loading dataset "mpg" from the ```load_dataset``` method in seaborn module. The goal will be to use linear regression to predict mpg - miles per gallon. 

&nbsp; a) Start by doing some initial EDA such as info(), describe() and figure out what you want to do with the missing values.

&nbsp; b) Use describe only on those columns that are relevant to get statistical information from. 

&nbsp; c) Make some plots on some of the columns that you find interesting.

&nbsp; d) Check if there are any columns you might want to drop. 

<details>

<summary>Answer</summary>

a) I have chosen to drop the rows, but it doesn't neccessary have to be the best method. Maybe some NaNs should be filled somehow?

b)
|      |      mpg |   cylinders |   displacement |   horsepower |   weight |   acceleration |
|:-----|---------:|------------:|---------------:|-------------:|---------:|---------------:|
| mean | 23.4459  |     5.47194 |        194.412 |     104.469  | 2977.58  |       15.5413  |
| std  |  7.80501 |     1.70578 |        104.644 |      38.4912 |  849.403 |        2.75886 |
| min  |  9       |     3       |         68     |      46      | 1613     |        8       |
| 25%  | 17       |     4       |        105     |      75      | 2225.25  |       13.775   |
| 50%  | 22.75    |     4       |        151     |      93.5    | 2803.5   |       15.5     |
| 75%  | 29       |     8       |        275.75  |     126      | 3614.75  |       17.025   |
| max  | 46.6     |     8       |        455     |     230      | 5140     |       24.8     |


c) Here are some example plots

<img src="../assets/EDA_mpg.png" height="400"/>

d) I have chosen to drop the columns origin and name. Think yourself if it is reasonable and feel free to experiment. Also there might be some domain experts in our class, that you can ask. 

</details>

---

In [725]:
import pandas as pd
import numpy as np
import seaborn as sns

df = sns.load_dataset("mpg")
df = df.reset_index()
df.head(10)

index   mpg  cylinders  displacement  horsepower  weight  acceleration  \
0      0  18.0          8         307.0       130.0    3504          12.0   
1      1  15.0          8         350.0       165.0    3693          11.5   
2      2  18.0          8         318.0       150.0    3436          11.0   
3      3  16.0          8         304.0       150.0    3433          12.0   
4      4  17.0          8         302.0       140.0    3449          10.5   
5      5  15.0          8         429.0       198.0    4341          10.0   
6      6  14.0          8         454.0       220.0    4354           9.0   
7      7  14.0          8         440.0       215.0    4312           8.5   
8      8  14.0          8         455.0       225.0    4425          10.0   
9      9  15.0          8         390.0       190.0    3850           8.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino  
5          70    usa           ford galaxie 500  
6          70    usa           chevrolet impala  
7          70    usa          plymouth fury iii  
8          70    usa           pontiac catalina  
9          70    usa         amc ambassador dpl

In [726]:
df.describe()

index         mpg   cylinders  displacement  horsepower  \
count  398.000000  398.000000  398.000000    398.000000  392.000000   
mean   198.500000   23.514573    5.454774    193.425879  104.469388   
std    115.036951    7.815984    1.701004    104.269838   38.491160   
min      0.000000    9.000000    3.000000     68.000000   46.000000   
25%     99.250000   17.500000    4.000000    104.250000   75.000000   
50%    198.500000   23.000000    4.000000    148.500000   93.500000   
75%    297.750000   29.000000    8.000000    262.000000  126.000000   
max    397.000000   46.600000    8.000000    455.000000  230.000000   

            weight  acceleration  model_year  
count   398.000000    398.000000  398.000000  
mean   2970.424623     15.568090   76.010050  
std     846.841774      2.757689    3.697627  
min    1613.000000      8.000000   70.000000  
25%    2223.750000     13.825000   73.000000  
50%    2803.500000     15.500000   76.000000  
75%    3608.000000     17.175000   79.000000  
max    5140.000000     24.800000   82.000000

In [736]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    float64
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    float64
 5   acceleration  392 non-null    float64
 6   model_year    392 non-null    float64
 7   origin        392 non-null    object 
 8   name          392 non-null    object 
dtypes: float64(7), object(2)
memory usage: 30.6+ KB


In [727]:
datatype = {"quant": ["mpg", "cylinders", "displacement", "horsepower", "weight", "acceleration", "model_year"], 
            "qual": ["origin", "name"]}

quants = df[datatype["quant"]].astype(np.float_)
df = pd.concat([quants, df[datatype["qual"]]], axis=1)
df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0        8.0         307.0       130.0  3504.0          12.0   
1    15.0        8.0         350.0       165.0  3693.0          11.5   
2    18.0        8.0         318.0       150.0  3436.0          11.0   
3    16.0        8.0         304.0       150.0  3433.0          12.0   
4    17.0        8.0         302.0       140.0  3449.0          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0        4.0         140.0        86.0  2790.0          15.6   
394  44.0        4.0          97.0        52.0  2130.0          24.6   
395  32.0        4.0         135.0        84.0  2295.0          11.6   
396  28.0        4.0         120.0        79.0  2625.0          18.6   
397  31.0        4.0         119.0        82.0  2720.0          19.4   

     model_year  origin                       name  
0          70.0     usa  chevrolet chevelle malibu  
1          70.0     usa          buick skylark 320  
2          70.0     usa         plymouth satellite  
3          70.0     usa              amc rebel sst  
4          70.0     usa                ford torino  
..          ...     ...                        ...  
393        82.0     usa            ford mustang gl  
394        82.0  europe                  vw pickup  
395        82.0     usa              dodge rampage  
396        82.0     usa                ford ranger  
397        82.0     usa                 chevy s-10  

[398 rows x 9 columns]

In [728]:
df.dropna(inplace=True)
X = df.drop(columns=["origin","name"])
# df.corr()["mpg"]

In [729]:
X, y = X.drop("mpg", axis=1), df["mpg"]
X.insert(0, "Intercept", 1)
y.shape

(392,)

## 1. Train|test split (*)

We want to predict the "mpg", split up X and y, and perform train|test split using scikit-learn. Choose test_size of 0.2 and random_state 42. Control the shapes of each X_train, X_test, y_train, y_test.  

<details>

<summary>Answer</summary>

Do a manual calculation to check against the shapes after train|test split. 

</details>

---

In [730]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size=0.2)

# intercept_const = pd.DataFrame({"intercept": np.ones(df.shape[0])})

# X = pd.concat([intercept_const, df["horsepower"]], axis=1)
# y = df["mpg"]
X_train.shape, X_test.shape, y_train.shape, y_test.shape, X_train

((313, 7),
 (79, 7),
 (313,),
 (79,),
      Intercept  cylinders  displacement  horsepower  weight  acceleration  \
 260          1        6.0         225.0       110.0  3620.0          18.7   
 184          1        4.0         140.0        92.0  2572.0          14.9   
 174          1        6.0         171.0        97.0  2984.0          14.5   
 64           1        8.0         318.0       150.0  4135.0          13.5   
 344          1        4.0          86.0        64.0  1875.0          16.4   
 ..         ...        ...           ...         ...     ...           ...   
 72           1        8.0         304.0       150.0  3892.0          12.5   
 107          1        6.0         232.0       100.0  2789.0          15.0   
 272          1        4.0         151.0        85.0  2855.0          17.6   
 352          1        4.0          98.0        65.0  2380.0          20.7   
 103          1        8.0         400.0       150.0  4997.0          14.0   
 
      model_year  
 260 

## 2. Function for evaluation (*)

Create a function for training a regression model, predicting and computing the metrics MAE, MSE, RMSE. It should take in parameters of X_train, X_test, y_train, y_test, model. Now create a linear regression model using scikit-learns ```LinearRegression()``` (OLS normal equation with SVD) and call your function to get metrics. 

<details>

<summary>Answer</summary>

MAE 2.50

MSE 10.50

RMSE 3.24

</details>

In [731]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, root_mean_squared_error

def training_model(X_train, X_test, y_train, y_test):
    model = LinearRegression()
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)
    
    MAE = mean_absolute_error(y_test, y_pred)
    MSE = mean_squared_error(y_test, y_pred)
    RMSE = root_mean_squared_error(y_test, y_pred)
    
    return MAE, MSE, RMSE

lr = training_model(X_train, X_test, y_train, y_test)
print(lr)

(2.5038600897761305, 10.502370329417333, 3.2407360783342622)


---
## 3. Compare models (*)

Create the following models 
- Linear regression (SVD)
- Linear regression (SVD) with scaled data (feature standardization)
- Polynomial linear regression with degree 1
- Polynomial linear regression with degree 2
- Polynomial linear regression with degree 3

Make a DataFrame with evaluation metrics and model. Which model performed overall best?

<details>

<summary>Answer</summary>

|      |   Linear regr. SVD |   Linear regr. SVD scaled |   Linear regr. SGD |   Polynom. regr. deg 1 |   Polynom. regr. deg 2 |   Polynom. regr. deg 3 |
|:-----|-------------------:|--------------------------:|-------------------:|-----------------------:|-----------------------:|-----------------------:|
| mae  |            2.50386 |                   2.50386 |            2.53515 |                2.50386 |                1.98048 |                2.11788 |
| mse  |           10.5024  |                  10.5024  |           10.8908  |               10.5024  |                7.41986 |                9.27353 |
| rmse |            3.24074 |                   3.24074 |            3.30012 |                3.24074 |                2.72394 |                3.04525 |

</details>

---

In [732]:
from sklearn.preprocessing import StandardScaler

""" Genom att skala data innan modellen tränas, kan du förbättra modellens prestanda 
och minska risken för överfitting. """

def scaled_training_model(X_train, X_test, y_train, y_test):
    scaler = StandardScaler()
    scaled_X_train = scaler.fit_transform(X_train)
    scaled_X_test = scaler.transform(X_test)

    model = LinearRegression()
    model.fit(scaled_X_train, y_train)

    y_pred = model.predict(scaled_X_test)
    
    scaled_MAE = mean_absolute_error(y_test, y_pred)
    scaled_MSE = mean_squared_error(y_test, y_pred)
    scaled_RMSE = root_mean_squared_error(y_test, y_pred)
    
    return scaled_MAE, scaled_MSE, scaled_RMSE

scaled_lr = scaled_training_model(X_train, X_test, y_train, y_test)

df_results = pd.DataFrame(scaled_lr)
df_results.insert(0, "", ["MAE", "MSE", "RMSE"])
df_results.rename(columns={df_results.columns[1]: "LR scaled"}, inplace=True)
df_results.insert(1, "LR", [lr[0], lr[1], lr[2]])  

In [733]:
from sklearn.preprocessing import PolynomialFeatures

def polynomial_regression(X_train, X_test, y_train, y_test):
    results = []
    deg = (1, 2, 3)
    for i in deg:
        poly_features = PolynomialFeatures(degree=i)
        poly_X_train = poly_features.fit_transform(X_train)
        poly_X_test = poly_features.transform(X_test)

        model = LinearRegression()
        model.fit(poly_X_train, y_train)

        y_pred = model.predict(poly_X_test)
        
        poly_MAE = mean_absolute_error(y_test, y_pred)
        poly_MSE = mean_squared_error(y_test, y_pred)
        poly_RMSE = root_mean_squared_error(y_test, y_pred)
        
        results.extend([[poly_MAE, poly_MSE, poly_RMSE]])  

    return results

poly = polynomial_regression(X_train, X_test, y_train, y_test)

for i, result in enumerate(poly):
    df_results.insert(3, f"LR Poly {3 - i}", poly[2-i])

df_results

LR  LR scaled  LR Poly 1  LR Poly 2  LR Poly 3
0   MAE   2.503860   2.503860   2.503860   1.980477   2.016261
1   MSE  10.502370  10.502370  10.502370   7.419858   8.114671
2  RMSE   3.240736   3.240736   3.240736   2.723942   2.848626


## 4. Further explorations (**)

Feel free to further explore the dataset, for example you could choose to 
- drop different columns
- find out feature importance in polynomial models
- fine tune further for a specific model by exploring hyperparameters (check documentation which type of parameters that can be changed)

---

Kokchun Giang

[LinkedIn][linkedIn_kokchun]

[GitHub portfolio][github_portfolio]

[linkedIn_kokchun]: https://www.linkedin.com/in/kokchungiang/
[github_portfolio]: https://github.com/kokchun/Portfolio-Kokchun-Giang

---